core/src/main/java/org/infinispan/interceptors/distribution/BaseDistributionInterceptor.java

```diff
@@ -504,15 +511,63 @@ public Object visitGetAllCommand(InvocationContext ctx, GetAllCommand command) t
                }
             }
          }
-         Map<Object, Object> values = (Map<Object, Object>) invokeNextInterceptor(ctx, command);
-         return values;
+         return invokeNextInterceptor(ctx, command);
+      }
+   }
+
+   protected void wrapInternalCacheEntry(InternalCacheEntry ice, InvocationContext ctx, Object key,
+                                         boolean isWrite, FlagAffectedCommand command) {
+      if (!ctx.replaceValue(key, ice))  {
+         if (isWrite)
+            entryFactory.wrapEntryForPut(ctx, key, ice, false, command, true);
+         else
+            ctx.putLookedUpEntry(key, ice);
       }
    }
 
    /**
     * @return Whether a remote get is needed to obtain the previous values of the affected entries.
     */
-   protected abstract boolean needValuesFromPreviousOwners(InvocationContext ctx, WriteCommand command);
+   protected abstract boolean writeNeedsRemoteValue(InvocationContext ctx, WriteCommand command, Object key);
+
+   protected boolean valueIsMissing(CacheEntry entry) {
+      return entry == null || (entry.isNull() && !entry.isRemoved() && !entry.skipLookup());
+   }
+
+   protected abstract boolean previousValueIsRemote(Object key);
 
    protected abstract void remoteGetBeforeWrite(InvocationContext ctx, WriteCommand command, Object key) throws Throwable;
+
+   protected boolean readNeedsRemoteGet(InvocationContext ctx, AbstractDataCommand command) {
+      if (command.hasFlag(Flag.CACHE_MODE_LOCAL) || command.hasFlag(Flag.SKIP_REMOTE_LOOKUP)) {
+         return false;
+      }
+      Object key = command.getKey();
+      ConsistentHash ch = stateTransferManager.getCacheTopology().getReadConsistentHash();
```

Comment:
This is not always used.  Maybe we should do ``!isValueAvailableLocally(stateTransferManager.getCacheTopology().getReadConsistentHash(), key);`` and assign it in the if block below?